File: storio-sqlite-annotation-processor/src/main/java/com/pushtorefresh/storio/sqlite/processor/generate/GetResolverGenerator.java
package com.pushtorefresh.storio.sqlite.processor.generate;

import com.pushtorefresh.storio.sqlite.processor.ProcessingException;
import com.pushtorefresh.storio.sqlite.processor.introspection.JavaType;
import com.pushtorefresh.storio.sqlite.processor.introspection.StorIOSQLiteColumnMeta;
import com.pushtorefresh.storio.sqlite.processor.introspection.StorIOSQLiteTypeMeta;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec;

import org.jetbrains.annotations.NotNull;

import static com.pushtorefresh.storio.sqlite.processor.generate.Common.ANDROID_NON_NULL_ANNOTATION_CLASS_NAME;
import static com.pushtorefresh.storio.sqlite.processor.generate.Common.INDENT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.BOOLEAN;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.BOOLEAN_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.DOUBLE;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.DOUBLE_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.FLOAT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.FLOAT_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.INTEGER;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.INTEGER_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.LONG;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.LONG_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.SHORT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.SHORT_OBJECT;
import static com.pushtorefresh.storio.sqlite.processor.introspection.JavaType.STRING;
import static javax.lang.model.element.Modifier.PUBLIC;

public class GetResolverGenerator {

    @NotNull
    public JavaFile generateJavaFile(@NotNull StorIOSQLiteTypeMeta storIOSQLiteTypeMeta) {
        final ClassName storIOSQLiteTypeClassName = ClassName.get(storIOSQLiteTypeMeta.packageName, storIOSQLiteTypeMeta.simpleName);

        final TypeSpec getResolver = TypeSpec.classBuilder(storIOSQLiteTypeMeta.simpleName + "StorIOSQLiteGetResolver")
                .addJavadoc("Generated resolver for Get Operation\n")
                .addModifiers(PUBLIC)
                .superclass(ParameterizedTypeName.get(ClassName.get("com.pushtorefresh.storio.sqlite.operation.get", "DefaultGetResolver"), storIOSQLiteTypeClassName))
                .addMethod(createMapFromCursorMethodSpec(storIOSQLiteTypeMeta, storIOSQLiteTypeClassName))
                .build();

        return JavaFile
                .builder(storIOSQLiteTypeMeta.packageName, getResolver)
                .indent(INDENT)
                .build();
    }

    @NotNull
    MethodSpec createMapFromCursorMethodSpec(@NotNull StorIOSQLiteTypeMeta storIOSQLiteTypeMeta, @NotNull ClassName storIOSQLiteTypeClassName) {
        final MethodSpec.Builder builder = MethodSpec.methodBuilder("mapFromCursor")
                .addJavadoc("{@inheritDoc}\n")
                .addAnnotation(Override.class)
                .addAnnotation(ANDROID_NON_NULL_ANNOTATION_CLASS_NAME)
                .addModifiers(PUBLIC)
                .returns(storIOSQLiteTypeClassName)
                .addParameter(ParameterSpec.builder(ClassName.get("android.database", "Cursor"), "cursor")
                        .addAnnotation(ANDROID_NON_NULL_ANNOTATION_CLASS_NAME)
                        .build())
                .addStatement("$T object = new $T()", storIOSQLiteTypeClassName, storIOSQLiteTypeClassName)
                .addCode("\n");

        for (final StorIOSQLiteColumnMeta columnMeta : storIOSQLiteTypeMeta.columns.values()) {
            final String columnIndex = "cursor.getColumnIndex(\"" + columnMeta.storIOSQLiteColumn.name() + "\")";

            final String getFromCursor;

            final JavaType javaType = columnMeta.javaType;

            if (javaType == BOOLEAN || javaType == BOOLEAN_OBJECT) {
                getFromCursor = "getInt(" + columnIndex + ") == 1";
            } else if (javaType == SHORT || javaType == SHORT_OBJECT) {
                getFromCursor = "getShort(" + columnIndex + ")";
            } else if (javaType == INTEGER || javaType == INTEGER_OBJECT) {
                getFromCursor = "getInt(" + columnIndex + ")";
            } else if (javaType == LONG || javaType == LONG_OBJECT) {
                getFromCursor = "getLong(" + columnIndex + ")";
            } else if (javaType == FLOAT || javaType == FLOAT_OBJECT) {
                getFromCursor = "getFloat(" + columnIndex + ")";
            } else if (javaType == DOUBLE || javaType == DOUBLE_OBJECT) {
                getFromCursor = "getDouble(" + columnIndex + ")";
            } else if (javaType == STRING) {
                getFromCursor = "getString(" + columnIndex + ")";
            } else {
                throw new ProcessingException(columnMeta.element, "Can not generate GetResolver for field");
            }

            builder
                    .addStatement("object.$L = cursor.$L", columnMeta.fieldName, getFromCursor);
        }

        return builder
                .addCode("\n")
                .addStatement("return object")
                .build();
    }
}

File: storio-sqlite-annotations-processor/src/main/kotlin/com/pushtorefresh/storio3/sqlite/annotations/processor/generate/GetResolverGenerator.kt
package com.pushtorefresh.storio3.sqlite.annotations.processor.generate

import com.pushtorefresh.storio3.common.annotations.processor.generate.Common.INDENT
import com.pushtorefresh.storio3.common.annotations.processor.generate.Common.getFromCursorString
import com.pushtorefresh.storio3.common.annotations.processor.generate.Generator
import com.pushtorefresh.storio3.sqlite.annotations.processor.introspection.StorIOSQLiteTypeMeta
import com.squareup.javapoet.*
import javax.lang.model.element.ElementKind
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.Modifier.PUBLIC

private const val SUFFIX = "StorIOSQLiteGetResolver"

object GetResolverGenerator : Generator<StorIOSQLiteTypeMeta> {

    override fun generateJavaFile(typeMeta: StorIOSQLiteTypeMeta): JavaFile {
        val className = ClassName.get(typeMeta.packageName, typeMeta.simpleName)

        val getResolver = TypeSpec.classBuilder(generateName(typeMeta))
                .addJavadoc("Generated resolver for Get Operation.\n")
                .addModifiers(PUBLIC)
                .superclass(ParameterizedTypeName.get(ClassName.get("com.pushtorefresh.storio3.sqlite.operations.get", "DefaultGetResolver"), className))
                .addMethod(
                        if (typeMeta.needsCreator)
                            createMapFromCursorWithCreatorMethodSpec(typeMeta, className)
                        else
                            createMapFromCursorMethodSpec(typeMeta, className))
                .build()

        return JavaFile
                .builder(typeMeta.packageName, getResolver)
                .indent(INDENT)
                .build()
    }

    private fun createMapFromCursorMethodSpec(typeMeta: StorIOSQLiteTypeMeta, className: ClassName): MethodSpec {
        val builder = MethodSpec.methodBuilder("mapFromCursor")
                .addJavadoc("{@inheritDoc}\n")
                .addAnnotation(Override::class.java)
                .addAnnotation(typeMeta.nonNullAnnotationClass)
                .addModifiers(PUBLIC)
                .returns(className)
                .addParameter(ParameterSpec.builder(ClassName.get("com.pushtorefresh.storio3.sqlite", "StorIOSQLite"), "storIOSQLite")
                        .addAnnotation(typeMeta.nonNullAnnotationClass)
                        .build())
                .addParameter(ParameterSpec.builder(ClassName.get("android.database", "Cursor"), "cursor")
                        .addAnnotation(typeMeta.nonNullAnnotationClass)
                        .build())
                .addStatement("\$T object = new \$T()", className, className)
                .addCode("\n")

        typeMeta.columns.values.forEach { columnMeta ->
            val columnIndex = "cursor.getColumnIndex(\"${columnMeta.storIOColumn.name}\")"

            val javaType = columnMeta.javaType

            val getFromCursor = getFromCursorString(javaType, columnIndex)

            val isBoxed = javaType.isBoxedType
            // otherwise -> if primitive and value from cursor null -> fail early
            if (isBoxed) builder.beginControlFlow("if (!cursor.isNull(\$L))", columnIndex)

            builder.addStatement("object.\$L = cursor.\$L", columnMeta.elementName, getFromCursor)

            if (isBoxed) builder.endControlFlow()
        }

        return builder
                .addCode("\n")
                .addStatement("return object")
                .build()
    }

    private fun createMapFromCursorWithCreatorMethodSpec(typeMeta: StorIOSQLiteTypeMeta, className: ClassName): MethodSpec {
        val builder = MethodSpec.methodBuilder("mapFromCursor")
                .addJavadoc("{@inheritDoc}\n")
                .addAnnotation(Override::class.java)
                .addAnnotation(typeMeta.nonNullAnnotationClass)
                .addModifiers(PUBLIC)
                .returns(className)
                .addParameter(ParameterSpec.builder(ClassName.get("com.pushtorefresh.storio3.sqlite", "StorIOSQLite"), "storIOSQLite")
                        .addAnnotation(typeMeta.nonNullAnnotationClass)
                        .build())
                .addParameter(ParameterSpec.builder(ClassName.get("android.database", "Cursor"), "cursor")
                        .addAnnotation(typeMeta.nonNullAnnotationClass)
                        .build())
                .addCode("\n")

        val paramsBuilder = StringBuilder().apply { append("(") }
        var first = true
        typeMeta.orderedColumns.forEach { columnMeta ->
            val columnIndex = "cursor.getColumnIndex(\"${columnMeta.storIOColumn.name}\")"

            val javaType = columnMeta.javaType

            val getFromCursor = getFromCursorString(javaType, columnIndex)

            val name = TypeName.get(
                    if (columnMeta.element is ExecutableElement)
                        columnMeta.element.returnType
                    else
                        columnMeta.element.asType()
            )

            val isBoxed = javaType.isBoxedType
            if (isBoxed) { // otherwise -> if primitive and value from cursor null -> fail early
                builder.addStatement("\$T \$L = null", name, columnMeta.realElementName)
                builder.beginControlFlow("if (!cursor.isNull(\$L))", columnIndex)
                builder.addStatement("\$L = cursor.\$L", columnMeta.realElementName, getFromCursor)
                builder.endControlFlow()
            } else {
                builder.addStatement("\$T \$L = cursor.\$L", name, columnMeta.realElementName, getFromCursor)
            }

            if (!first) paramsBuilder.append(", ")
            first = false
            paramsBuilder.append(columnMeta.realElementName)
        }
        paramsBuilder.append(")")
        builder.addCode("\n")

        // creator can't be null here
        if (typeMeta.creator!!.kind == ElementKind.CONSTRUCTOR) {
            builder.addStatement("\$T object = new \$T$paramsBuilder", className, className)
        } else {
            builder.addStatement("\$T object = \$T.\$L", className, className, "${typeMeta.creator!!.simpleName}$paramsBuilder")
        }

        return builder
                .addCode("\n")
                .addStatement("return object")
                .build()
    }

    fun generateName(typeMeta: StorIOSQLiteTypeMeta) = "${typeMeta.simpleName}$SUFFIX"
}
